### PR TITLE
feat!: Deny replying on replies

### LINKED ISSUE #2674: Deny abilities to send replies on reply
### File Location(s)

_No response_

### Proposal

part of messaging proposal #invaria

### PR DESCRIPTION
**Pull request by breathx**
Resolves #2674.

TODO:
- [x] add error
- [x] deny syscalls
- [x] don't insert replies into mailbox
- [x] remove reply_for_reply etc api from gstd and gcore
- [x] fix demos and tests
- [x] remove limitations on gas limit for reply from syscall
- [x] adjust error replies
- [x] fix tests one more time
- [x] WasmProgram trait: handle_reply
- [x] write tests:
      - [x] that reply banned in handle_reply
      - [x] ~~reply never added into Mailbox~~ (indirectly tested)
      - [x] reply can be sent with 1 gas
     
- [x] rebase and fix issues after #2547

### CODE REVIEW

File: core-processor/src/executor.rs [hunk 1]
```diff
@@ -511,18 +506,21 @@ where
...
-                Default::default(),
-                Default::default(),
-                payload
-                    .try_into()
-                    .map_err(|e| format!("Failed to create payload: {e:?}"))?,
-                gas_limit,
-                Default::default(),
-                Default::default(),
+            IncomingDispatch::new(
+                DispatchKind::Handle,
```
Comment (NikVolf) [inline core-processor/src/executor.rs:510]: how is this not `DispatchKind::Reply` ?

File: pallets/gear/src/tests.rs [hunk 2]
```diff
@@ -10178,6 +10162,28 @@ mod utils {
...
+            }) = e.event
+            {
+                total += 1;
+                if expiration.is_some() {
+                    to_mailbox += 1;
+                }
+            };
+        });
+
+        (total, to_mailbox)
```
Comment (gshep) [inline pallets/gear/src/tests.rs:10184]:
```suggestion
        System::events().into_iter().fold((0usize, 0usize), |(total, to_mailbox), e| {
            if let MockRuntimeEvent::Gear(Event::UserMessageSent {
                expiration,
                ..
            }) = e.event
            {
                (total + 1, to_mailbox + if expiration.is_some() { 1 } else { 0 })
            } else {
                (total, to_mailbox)
            }
        })
```

File: gtest/src/manager.rs [hunk 3]
```diff
@@ -666,49 +666,22 @@ impl ExtManager {
         let response = match dispatch.kind() {
             DispatchKind::Init => mock.init(payload).map(Mocked::Reply),
             DispatchKind::Handle => mock.handle(payload).map(Mocked::Reply),
-            DispatchKind::Reply => mock.handle_reply(payload).map(Mocked::Reply),
+            DispatchKind::Reply => mock.handle_reply(payload).map(|()| Mocked::Reply(None)),
```
Comment (gshep) [inline gtest/src/manager.rs:669]: it is more common to use underscore

File: pallets/gear/src/tests.rs [hunk 4]
```diff
@@ -4060,10 +4051,21 @@ fn wake_messages_after_program_inited() {
...
         run_to_block(20, None);
 
-        let actual_n = MailboxOf::<Test>::iter_key(USER_3).fold(0usize, |i, (m, _bn)| {
-            assert_eq!(m.payload().to_vec(), b"Hello, world!".encode());
-            i + 1
-        });
+        let actual_n = System::events()
+            .into_iter()
+            .filter_map(|e| {
+                if let MockRuntimeEvent::Gear(Event::UserMessageSent { message, .. }) = e.event {
```
Comment (gshep) [inline pallets/gear/src/tests.rs:4057]: wouldn't the usual `match` expression fit better there?

File: pallets/gear/src/tests.rs [hunk 5]
```diff
@@ -10642,17 +10745,24 @@ fn relay_messages() {
...
+            System::events().into_iter().for_each(|e| {
+                if let MockRuntimeEvent::Gear(Event::UserMessageSent { message, .. }) = e.event {
+                    let Expected { user, payload } = &expected[i];
+
+                    if message.destination().into_origin() == user.into_origin() {
+                        assert_eq!(message.payload(), payload, "{label}");
+                        i += 1;
+                    }
+                }
+            });
```
Comment (gshep) [inline pallets/gear/src/tests.rs:10763]: `fold` may fit better there doesn't it?

File: core-processor/src/executor.rs (hunk 1 above)
Comment (breathx) [reply core-processor/src/executor.rs:510]: execute_for_reply fn used to read state of wasms
it doesn't execute reply, but executes custom entry to collect error or only reply sent
so we need to throw here any kind that will allow replying (Init or Handle)

**Pull request merged**